source/common/config/decoded_resource_impl.h

```diff
@@ -20,58 +21,95 @@ repeatedPtrFieldToVector(const Protobuf::RepeatedPtrField<std::string>& xs) {
 
 } // namespace
 
+class DecodedResourceImpl;
+using DecodedResourceImplPtr = std::unique_ptr<DecodedResourceImpl>;
+
 class DecodedResourceImpl : public DecodedResource {
 public:
-  DecodedResourceImpl(OpaqueResourceDecoder& resource_decoder, const ProtobufWkt::Any& resource,
-                      const std::string& version)
-      : DecodedResourceImpl(resource_decoder, {}, Protobuf::RepeatedPtrField<std::string>(),
-                            resource, true, version) {}
+  static DecodedResourceImpl maybeUnwrap(OpaqueResourceDecoder& resource_decoder,
+                                         const ProtobufWkt::Any& resource,
+                                         const std::string& version) {
+    if (resource.Is<envoy::service::discovery::v3::Resource>()) {
+      envoy::service::discovery::v3::Resource r;
+      MessageUtil::unpackTo(resource, r);
+
+      r.set_version(version);
+
+      return DecodedResourceImpl(resource_decoder, r);
+    }
+
+    return DecodedResourceImpl(resource_decoder, {}, Protobuf::RepeatedPtrField<std::string>(),
+                               resource, true, version, absl::nullopt);
+  }
+
+  static DecodedResourceImplPtr maybeUnwrapPtr(OpaqueResourceDecoder& resource_decoder,
```

Comment:
Can we call this something like `fromResource`? The unwrapping is an internal implementation detail.